src-extras/Database/LSMTree/Extras/UTxO.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE NamedFieldPuns             #-}
+
+module Database.LSMTree.Extras.UTxO (
+    UTxOKey (..)
+  , UTxOValue (..)
+  , UTxOBlob (..)
+  ) where
+
+import           Control.DeepSeq
+import qualified Data.ByteString as BS
+import qualified Data.Primitive as P
+import           Data.WideWord.Word256
+import           Data.Word
+import           Database.LSMTree.Extras.Orphans ()
+import qualified Database.LSMTree.Internal.RawBytes as RB
+import           Database.LSMTree.Internal.Serialise
+import           Database.LSMTree.Internal.Serialise.Class as Class
+import           Database.LSMTree.Internal.Vector
+import           GHC.Generics
+import           System.Random
+
+{-------------------------------------------------------------------------------
+  UTxO keys
+-------------------------------------------------------------------------------}
+
+-- | A model of a UTxO key after @TxIn@: a 256-bit hash, 64-bit identifier
+data UTxOKey = UTxOKey {
+    txId :: !Word256 -- no unpack, since the @TxId@ field doesn't have it
+  , txIx :: {-# UNPACK #-} !Word64
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving anyclass (Uniform, NFData)
+
+instance SerialiseKey UTxOKey where
+  serialiseKey UTxOKey{txId, txIx} =
+    RB.RawBytes $ mkPrimVector 0 40 $ P.runByteArray $ do
+      ba <- P.newByteArray 40
+      P.writeByteArray ba 0 $ byteSwap64 (word256hi txId)
+      P.writeByteArray ba 1 $ byteSwap64 (word256m1 txId)
+      P.writeByteArray ba 2 $ byteSwap64 (word256m0 txId)
+      P.writeByteArray ba 3 $ byteSwap64 (word256lo txId)
+      P.writeByteArray ba 4 $ byteSwap64 txIx
+      return ba
+  deserialiseKey = error "deserialiseKey: UTxOKey" -- TODO
+
+{-------------------------------------------------------------------------------
+  UTxO values
+-------------------------------------------------------------------------------}
+
+-- | A model of a UTxO value (512-bit)
+data UTxOValue = UTxOValue {
+    utxoValueHigh :: !Word256
+  , utxoValueLow  :: !Word256
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving anyclass (Uniform, NFData)
+
+instance SerialiseValue UTxOValue where
+  serialiseValue (UTxOValue hi lo) = Class.serialiseValue lo <> Class.serialiseValue hi
+  deserialiseValue = error "deserialiseValue: UTxOValue"
+  deserialiseValueN = error "deserialiseValueN: UTxOValue"
+
+{-------------------------------------------------------------------------------
+  UTxO blobs
+-------------------------------------------------------------------------------}
+
+-- | A blob of arbitrary size
+newtype UTxOBlob = UTxOBlob BS.ByteString
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving anyclass NFData
+
+instance SerialiseValue UTxOBlob where
+  serialiseValue (UTxOBlob bs) = Class.serialiseValue bs
+  deserialiseValue = error "deserialiseValue: UTxOBlob"
+  deserialiseValueN = error "deserialiseValueN: UTxOBlob"
```

Comment:
Do we have any idea for the size distribution of UTxO blobs, as well as how often they might appear? If so, we should have something here to generate them plus any relevant information. If not, maybe we just don't need a `UTxOBlob`, they are not mentioned in the requirements. We can test and benchmark blobs separately from the UTxO workload.

Reply:
I don't know, but we could find out when we work on the relevant work package. For now, I'm just picking this representation and some arbitrary distribution, but it's not very important if it's not super precise (yet)

Reply:
The `UTxOBlob` might still be useful for micro-benchmarks or tests, even if we don't use it in the macro-benchmark

Reply:
The point I have is, if we don't know what UTxO blobs in particular look like, why not just use any general-purpose `Blob` type? But we can leave it here with the idea of adding a somewhat representative generator later.